playground/backend/internal/code_processing/code_processing.go

```diff
@@ -0,0 +1,218 @@
+// At the end of this method deletes all created folders.
+func Process(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, compileBuilder *executors.CompileBuilder, pipelineId uuid.UUID, env *environment.Environment, sdk pb.Sdk) {
+	ctxWithTimeout, cancelByTimeoutFunc := context.WithTimeout(ctx, env.ApplicationEnvs.PipelineExecuteTimeout())
+	defer func(lc *fs_tool.LifeCycle) {
+		cancelByTimeoutFunc()
+		cleanUp(pipelineId, lc)
+	}(lc)
+
+	errorChannel := make(chan error, 1)
+	dataChannel := make(chan interface{}, 1)
+	doneChannel := make(chan bool, 1)
+	cancelChan := make(chan bool, 1)
+
+	go cancelCheck(ctx, pipelineId, cancelChan, cacheService)
+
+	// build executor for validate and compile steps
+	exec := compileBuilder.Build()
+
+	// validate
+	logger.Infof("%s: Validate() ...\n", pipelineId)
+	validateFunc := exec.Validate()
+	go validateFunc(doneChannel, errorChannel)
+
+	if !processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, doneChannel, nil, errorChannel, pb.Status_STATUS_VALIDATION_ERROR, pb.Status_STATUS_COMPILING) {
+		return
+	}
+
+	// compile
+	logger.Infof("%s: Compile() ...\n", pipelineId)
+	compileCmd := exec.Compile(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := compileCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
+
+	if !processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, doneChannel, dataChannel, errorChannel, pb.Status_STATUS_COMPILE_ERROR, pb.Status_STATUS_EXECUTING) {
+		return
+	}
+
+	runBuilder, err := setuper.SetupRunBuilder(pipelineId, lc, sdk, env, compileBuilder)
+	if err != nil {
+		logger.Errorf("%s: error during setuper runBuilder: %s\n", pipelineId, err.Error())
+		cacheService.SetValue(ctxWithTimeout, pipelineId, cache.Status, pb.Status_STATUS_ERROR)
+		return
+	}
+
+	// build executor for run step
+	exec = runBuilder.Build()
+
+	// run
+	logger.Infof("%s: Run() ...\n", pipelineId)
+	runCmd := exec.Run(ctxWithTimeout)
+	go func(doneCh chan bool, errCh chan error, dataCh chan interface{}) {
+		data, err := runCmd.CombinedOutput()
+		dataCh <- data
+		if err != nil {
+			errCh <- err
+			doneCh <- false
+		} else {
+			doneCh <- true
+		}
+	}(doneChannel, errorChannel, dataChannel)
+
+	processStep(ctxWithTimeout, pipelineId, cacheService, cancelChan, doneChannel, dataChannel, errorChannel, pb.Status_STATUS_RUN_ERROR, pb.Status_STATUS_FINISHED)
+}
+
+// processStep processes each executor's step with cancel and timeout checks.
+// If finishes by canceling, timeout or error - returns false.
+// If finishes successfully returns true.
+func processStep(ctx context.Context, pipelineId uuid.UUID, cacheService cache.Cache, cancelChan, doneChannel chan bool, dataChan chan interface{}, errorChannel chan error, errorCaseStatus, successCaseStatus pb.Status) bool {
```

Comment:
why you can't use errorCaseStatus, successCaseStatus pb.Status directly in function. these are global, aren't they?

Reply:
Yes, they are global. Now the `processStep()` is being called 3 times:
- Validation: this function receives `pb.Status_STATUS_VALIDATION_ERROR` and `pb.Status_STATUS_COMPILING`
- Compile: this function receives `pb.Status_STATUS_COMPILE_ERROR` and `pb.Status_STATUS_EXECUTING`
- Run: this function receives `pb.Status_STATUS_RUN_ERROR`, `pb.Status_STATUS_FINISHED`

So each case has his own `errorCaseStatus` and `successCaseStatus`.
Does it answer your question?